Add form validation to the woo installer address step

lsl opened the pull request:
#### Changes proposed in this Pull Request

* Add form validation to the woo installer's address step.

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Go to http://calypso.localhost:3000/woocommerce-installation/?flags=woop
* Choose a non atomic site (atomic won't work until JP updates with new country endpoint data)
* Click set up my store
* You should be taken to the address step (when the woop flag is active)
* Clicking the continue button should fail with validation errors until the form is completed correctly.
* Once completing the install and transfer flow you'll land on Woo's original setup profile, your address should be prepopulated using atomic data - though I noticed a timing issue with fetching country info (the form errored)

Before
![Screenshot 2022-01-13 at 13-21-47 Add WooCommerce to your site — WordPress com](https://user-images.githubusercontent.com/811776/149254460-9960f9e7-0d1e-4eb9-b863-5f6773ab40e0.png)

After
![Screenshot 2022-01-13 at 13-58-06 Add WooCommerce to your site — WordPress com](https://user-images.githubusercontent.com/811776/149257937-6800a1a6-f128-4c7b-9931-da1a668b465d.png)

Fixes #59857

lsl committed:
Basic required validation check

lsl committed:
try: FormInputValidation

lsl committed:
try: hook usage not triggering re-render after validation.. probably need to simplify this

allilevine committed:
Add email field error

a8ci18n commented on client/signup/steps/woocommerce-install/step-store-address/index.tsx:
```diff
@@ -162,3 +203,56 @@ export default function StepStoreAddress( props: WooCommerceInstallProps ): Reac
+		errors[ WOOCOMMERCE_ONBOARDING_PROFILE ] = ! get( WOOCOMMERCE_ONBOARDING_PROFILE )[
+			'store_email'
+		]
+			? __( 'Email address is required.' )
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **96** times:
`translate( 'A valid email address is required' )` *ES Score: 7*
See 1 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/woop-validation)

lsl committed:
Make error messages match woo, reuse translations

lsl committed:
hack: re-render on validation

lsl committed:
Reenable country reset, fix css

@obenland needed this for testing validation but I don't see any way to
reset the value without it so added it back for usability. Also fixed a
margin that didn't seem to match the text controls.

lsl committed:
Align error symbol and text

lsl committed:
Validate email address value

lsl committed:
Moved country below city/postcode

lsl committed:
Replaced "Region" with "State" for american usage.

This makes more sense for most users and can be localized
to "Region" as needed.

lsl committed:
Word email error the same as the others

lsl committed:
Style input boxes red on error

ramonjd commented on client/signup/steps/woocommerce-install/step-store-address/index.tsx:
```diff
@@ -162,3 +212,56 @@ export default function StepStoreAddress( props: WooCommerceInstallProps ): Reac
+		errors[ WOOCOMMERCE_ONBOARDING_PROFILE ] = ! emailValidator.validate(
+			get( WOOCOMMERCE_ONBOARDING_PROFILE )?.[ 'store_email' ]
+		)
+			? __( 'Please add a store email' )
```
Please ignore if not relevant. but is "store email" a thing? 

What about a more generic `Please add a valid email address`?

lsl replied:
Yea, I think you're right, I'll do this in a follow up. I think the error message positioning needs more tweaking anyway.

ramonjd reviewed (approved):
Works as described.

I don't have a load of context on the code, but the form validation works as expected, whereas prod as no validation.

I can continue onto the cart flow after adding valid input entries.

**Desktop**
<img width="400" alt="Screen Shot 2022-01-13 at 2 30 04 pm" src="https://user-images.githubusercontent.com/6458278/149261553-dc952f72-59fb-4dfb-92e7-86b2e03f133e.png">


**Mobile**
<img width="300" alt="Screen Shot 2022-01-13 at 2 29 58 pm" src="https://user-images.githubusercontent.com/6458278/149261563-b3204791-5d51-4e94-b875-eb813da686a5.png">

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7082003

Thank you @lsl for including a screenshot in the description! This is really helpful for our translators.

lsl commented on client/signup/steps/woocommerce-install/step-store-address/index.tsx:
```diff
@@ -75,6 +76,19 @@ export default function StepStoreAddress( props: WooCommerceInstallProps ): Reac
+	const [ submitted, setSubmitted ] = useState( 0 );
+	useEffect( () => {
+		return;
+	}, [ submitted ] );
```
@allilevine this is a bit of a hack but fixes the re-render problem when validation runs. This should really be cleaned up next sprint but I think it will have to do for now.

ramonjd replied:
What's the re-render problem? It looks like the hook will run every time the `submitted` dependency is updated and won't do anything.

It might be worth looking into wrapping validation functions in a [useCallback](https://reactjs.org/docs/hooks-reference.html#usecallback)... maybe?

allilevine replied:
The form wasn't re-rendering on submit, so the errors sometimes weren't displaying. I think by running the hook when `submittted` is updated, we're re-rendering on submit. But I would think that updating state (`submitted`) would be enough to re-render the component. 🤔

lsl commented on client/signup/steps/woocommerce-install/step-store-address/index.tsx:
```diff
@@ -83,49 +97,85 @@ export default function StepStoreAddress( props: WooCommerceInstallProps ): Reac
+						} }
+						className={ address1Error ? 'is-error' : '' }
 					/>
+					<ControlError error={ address1Error } />
```
For anyone that comes along, this is not how we should be doing error handling / validation. It sort of works but we really need a component that wraps the error and the text control together. What we might want to look at is updating the `form-input` components we already use in calypso to make use of the core component `TextControl` but that might not be doable. cc @SaxonF re library thinkings

lsl commented on client/signup/steps/woocommerce-install/step-store-address/index.tsx:
```diff
@@ -162,3 +212,56 @@ export default function StepStoreAddress( props: WooCommerceInstallProps ): Reac
+	return null;
+}
+
+function useAddressFormValidation( siteId: number ) {
```
I don't think using a hook for this makes sense, I was trying something out and it made its way into production. 

Ideally we clean this up next sprint. I think all this state should get moved up out of the site settings hook and the error validation gets coupled in with the values they refer to.

The site settings hook might still make sense but I think the form still needs to retain control over state/value changes. The hook should only be involved on submit. Or maybe it makes sense to move this validation code into the site settings hook somehow. :thinking:

lsl commented on client/signup/steps/woocommerce-install/step-store-address/style.scss:
```diff
@@ -2,11 +2,42 @@
+			border-color: var( --color-error );
+			box-shadow: 0 0 0 0 var( --color-error );
+		}
+	}
```
This css is mostly just unsetting css on the core components in an unsustainable way. Feels like we're using the wrong components for this or they're too opinionated or we're using them wrong. Maybe try looking at using calypso's `Form` components in the follow up? #60018

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7082003).
